src/server/multiplayer/MultiplayerServerState.js

```diff
@@ -152,4 +186,60 @@ export default class MultiplayerServerState {
   leaveGame = (clientID) => {
     this.state.games = compact(this.state.games.map(game => withoutClient(game, clientID)));
   }
+
+  // Handle end-of-game actions.
+  endGame = (game) => {
+    if (game.state.winner) {
+      this.storeGameResult(game);
+      // TODO Alter player ratings accordingly.
+    }
+  }
+
+  // Store the result of a game in Firebase.
+  storeGameResult = (game) => {
+    const { ids, format, type, state: { winner } } = game;
+    saveGame({
+      players: mapValues(ids, clientID => this.getClientUserData(clientID).uid),
+      format,
+      type,
+      winner
+    });
+  }
+
+  // Add a player to the matchmaking queue.
+  joinQueue = (clientID, deck) => {
+    this.state.matchmakingQueue.push({ clientID, deck });
+  }
+
+  // Remove a player from the matchmaking queue.
+  leaveQueue = (clientID) => {
+    this.state.matchmakingQueue = reject(this.state.matchmakingQueue, { clientID });
+  }
+
+  // Return pairs of player IDs to match into games.
+  // TODO: Fix this, using MMR.
+  findAvailableMatches = () => {
+    const playerIds = this.state.matchmakingQueue.map(m => m.clientID);
+    return chunk(playerIds, 2).filter(m => m.length === 2);
+  }
+
+  // Pair players if there are at least two people waiting for a ranked game.
+  // Return all games created.
+  matchPlayersIfPossible = () => {
+    const { matchmakingQueue } = this.state;
+    const playerPairs = this.findAvailableMatches();
+
+    return playerPairs.map(([pid1, pid2]) => {
+      const gameName = `Ranked#${this.getClientUsername(pid1)}-vs-${this.getClientUsername(pid2)}`;
```

Comment:
Again, why do we need the game type prefixed in the name if we also have a `type` field?

Reply:
see above